test/integration/integration_test.go

```diff
@@ -0,0 +1,208 @@
+		return user{}, err
+	}
+	responseBody, err := httpRequest("https://msidlab.com/api/user", query, accessToken)
+	if err != nil {
+		return user{}, err
+	}
+	var users []user
+	err = json.Unmarshal(responseBody, &users)
+	if err != nil {
+		return user{}, err
+	}
+	if len(users) == 0 {
+		return user{}, errors.New("No user found")
+	}
+	user := users[0]
+	user.Password, err = l.getSecret(map[string]string{"Secret": user.LabName})
+	if err != nil {
+		return user, err
+	}
+	return user, nil
+}
+
+func (l *labClient) getSecret(query map[string]string) (string, error) {
+	accessToken, err := l.getLabAccessToken()
+	if err != nil {
+		return "", err
+	}
+	responseBody, err := httpRequest("https://msidlab.com/api/LabUserSecret", query, accessToken)
+	if err != nil {
+		return "", err
+	}
+	var secret secret
+	err = json.Unmarshal(responseBody, &secret)
+	if err != nil {
+		return "", err
+	}
+	return secret.Value, nil
+}
+
+// TODO: Add getApp() when needed
+
+// TestConstants
+
+// Scopes.
+const (
+	msIDlabDefaultScope = "https://msidlab.com/.default"
+	graphDefaultScope   = "https://graph.windows.net/.default"
+)
+
+// MicrosoftAuthorityHost is the host authority for Microsoft.
+const microsoftAuthorityHost = "https://login.microsoftonline.com/"
+
+// Authority values.
+const (
+	organizationsAuthority = microsoftAuthorityHost + "organizations/"
+	microsoftAuthority     = microsoftAuthorityHost + "microsoft.onmicrosoft.com"
+	//msIDlabTenantAuthority = microsoftAuthorityHost + "msidlab4.onmicrosoft.com" - Will be needed in the furture
+)
+
+// Tests
+
+func getTestUser(lc *labClient, query map[string]string) user {
+	testUser, err := lc.getUser(query)
+	if err != nil {
+		panic("Failed to get input user. " + err.Error())
+	}
+	return testUser
+}
+
+func TestUsernamePassword(t *testing.T) {
+	labClientInstance, err := newLabClient()
+	if err != nil {
+		panic("Failed to get a lab client. " + err.Error())
+	}
+	tests := map[string]struct {
+		user user
+	}{
+		"Managed": {user: getTestUser(labClientInstance, map[string]string{"usertype": "cloud"})},
+		"ADFSv2":  {user: getTestUser(labClientInstance, map[string]string{"usertype": "federated", "federationProvider": "ADFSv2"})},
+		"ADFSv3":  {user: getTestUser(labClientInstance, map[string]string{"usertype": "federated", "federationProvider": "ADFSv3"})},
+		//"ADFSv4":  {user: getTestUser(labClientInstance, map[string]string{"usertype": "federated", "federationProvider": "ADFSv2"})},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			options := msal.DefaultPublicClientApplicationOptions()
+			options.Authority = organizationsAuthority
+			publicClientApp, err := msal.NewPublicClientApplication(tc.user.AppID, &options)
+			if err != nil {
+				panic(err)
+			}
+			result, err := publicClientApp.AcquireTokenByUsernamePassword(
+				context.Background(),
+				[]string{graphDefaultScope},
+				tc.user.Upn,
+				tc.user.Password,
+			)
+			if err != nil {
+				t.Error(err)
```

Comment:
This should be t.Fatalf(), because you don't want to proceed after this.  

This is where I usually do my desc field instead of doing sub tests:

```go
t.Fatalf("TestUsernamePassword(%s): on AcquireTokenByUsernamePassword(): got err == %s, want err == nil", test.desc, err)
```